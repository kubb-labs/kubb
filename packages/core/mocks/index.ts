import { pascalCase } from '../src/transformers/casing.ts'

import type { PluginManager } from '../src/PluginManager.ts'
import { readSync } from '../src/fs/read.ts'

export const mockedPluginManager = {
  resolveName: ({ name, type }) => {
    if (type === 'type') {
      return pascalCase(name)
    }

    return name
  },
  resolvePath: ({ baseName }) => baseName,
  logger: {
    emit(message) {
      console.log(message)
    },
    on(eventName, args) {},
    logLevel: 'info',
  },
  getFile: ({ name, extName, pluginKey }) => {
    const baseName = `${name}${extName}`
    let source = ''

    try {
      source = readSync(baseName)
    } catch (_e) {
      //
    }

    return {
      path: baseName,
      baseName,
      meta: {
        pluginKey,
      },
      source,
    }
  },
} as PluginManager
