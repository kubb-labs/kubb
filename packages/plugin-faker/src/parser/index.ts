import transformers from '@kubb/core/transformers'
import { SchemaGenerator, isKeyword, schemaKeywords } from '@kubb/plugin-oas'

import type { Schema, SchemaKeywordBase, SchemaKeywordMapper, SchemaMapper } from '@kubb/plugin-oas'
import type { Options } from '../types.ts'

export const fakerKeywordMapper = {
  any: () => 'undefined',
  unknown: () => 'unknown',
  number: (min?: number, max?: number) => {
    if (max !== undefined && min !== undefined) {
      return `faker.number.float({ min: ${min}, max: ${max} })`
    }

    if (min !== undefined) {
      return `faker.number.float({ min: ${min} })`
    }

    if (max !== undefined) {
      return `faker.number.float({ max: ${max} })`
    }

    return 'faker.number.float()'
  },
  integer: (min?: number, max?: number) => {
    if (max !== undefined && min !== undefined) {
      return `faker.number.int({ min: ${min}, max: ${max} })`
    }

    if (min !== undefined) {
      return `faker.number.int({ min: ${min} })`
    }

    if (max !== undefined) {
      return `faker.number.int({ max: ${max} })`
    }

    return 'faker.number.int()'
  },
  string: (min?: number, max?: number) => {
    if (max !== undefined && min !== undefined) {
      return `faker.string.alpha({ length: { min: ${min}, max: ${max} } })`
    }

    if (min !== undefined) {
      return `faker.string.alpha({ length: { min: ${min} } })`
    }

    if (max !== undefined) {
      return `faker.string.alpha({ length: { max: ${max} } })`
    }

    return 'faker.string.alpha()'
  },
  boolean: () => 'faker.datatype.boolean()',
  undefined: () => 'undefined',
  null: () => 'null',
  array: (items: string[] = []) => `faker.helpers.arrayElements([${items.join(', ')}]) as any`,
  tuple: (items: string[] = []) => `faker.helpers.arrayElements([${items.join(', ')}]) as any`,
  enum: (items: Array<string | number> = []) => `faker.helpers.arrayElement<any>([${items.join(', ')}])`,
  union: (items: string[] = []) => `faker.helpers.arrayElement<any>([${items.join(', ')}])`,
  /**
   * ISO 8601
   */
  datetime: () => 'faker.date.anytime().toISOString()',
  /**
   * Type `'date'` Date
   * Type `'string'` ISO date format (YYYY-MM-DD)
   * @default ISO date format (YYYY-MM-DD)
   */
  date: (type: 'date' | 'string' = 'string', parser: Options['dateParser'] = 'faker') => {
    if (type === 'string') {
      if (parser !== 'faker') {
        return `${parser}(faker.date.anytime()).format("YYYY-MM-DD")`
      }
      return 'faker.date.anytime().toString()'
    }

    if (parser !== 'faker') {
      throw new Error(`type '${type}' and parser '${parser}' can not work together`)
    }

    return 'faker.date.anytime()'
  },
  /**
   * Type `'date'` Date
   * Type `'string'` ISO time format (HH:mm:ss[.SSSSSS])
   * @default ISO time format (HH:mm:ss[.SSSSSS])
   */
  time: (type: 'date' | 'string' = 'string', parser: Options['dateParser'] = 'faker') => {
    if (type === 'string') {
      if (parser !== 'faker') {
        return `${parser}(faker.date.anytime()).format("HH:mm:ss")`
      }
      return 'faker.date.anytime().toString()'
    }

    if (parser !== 'faker') {
      throw new Error(`type '${type}' and parser '${parser}' can not work together`)
    }

    return 'faker.date.anytime()'
  },
  uuid: () => 'faker.string.uuid()',
  url: () => 'faker.internet.url()',
  and: (items: string[] = []) => `Object.assign({}, ${items.join(', ')})`,
  object: () => 'object',
  ref: () => 'ref',
  matches: (value = '', regexGenerator: 'faker' | 'randexp' = 'faker') => {
    if (regexGenerator === 'randexp') {
      return `${transformers.toRegExpString(value, 'RandExp')}.gen()`
    }
    return `faker.helpers.fromRegExp(${transformers.toRegExpString(value)})`
  },
  email: () => 'faker.internet.email()',
  firstName: () => 'faker.person.firstName()',
  lastName: () => 'faker.person.lastName()',
  password: () => 'faker.internet.password()',
  phone: () => 'faker.phone.number()',
  blob: () => 'faker.image.imageUrl() as unknown as Blob',
  default: undefined,
  describe: undefined,
  const: (value?: string | number) => (value as string) ?? '',
  max: undefined,
  min: undefined,
  nullable: undefined,
  nullish: undefined,
  optional: undefined,
  readOnly: undefined,
  strict: undefined,
  deprecated: undefined,
  example: undefined,
  schema: undefined,
  catchall: undefined,
  name: undefined,
} satisfies SchemaMapper<string | null | undefined>

/**
 * @link based on https://github.com/cellular/oazapfts/blob/7ba226ebb15374e8483cc53e7532f1663179a22c/src/codegen/generate.ts#L398
 */

function schemaKeywordsorter(a: Schema, b: Schema) {
  if (b.keyword === 'null') {
    return -1
  }

  return 0
}

export function joinItems(items: string[]): string {
  switch (items.length) {
    case 0:
      return 'undefined'
    case 1:
      return items[0]!
    default:
      return fakerKeywordMapper.union(items)
  }
}

type ParserOptions = {
  name: string
  typeName?: string
  description?: string

  seed?: number | number[]
  regexGenerator?: 'faker' | 'randexp'
  canOverride?: boolean
  dateParser?: Options['dateParser']
  mapper?: Record<string, string>
}

export function parse(parent: Schema | undefined, current: Schema, options: ParserOptions): string | null | undefined {
  const value = fakerKeywordMapper[current.keyword as keyof typeof fakerKeywordMapper]

  if (!value) {
    return undefined
  }

  if (isKeyword(current, schemaKeywords.union)) {
    return fakerKeywordMapper.union(current.args.map((schema) => parse(current, schema, { ...options, canOverride: false })).filter(Boolean))
  }

  if (isKeyword(current, schemaKeywords.and)) {
    return fakerKeywordMapper.and(current.args.map((schema) => parse(current, schema, { ...options, canOverride: false })).filter(Boolean))
  }

  if (isKeyword(current, schemaKeywords.array)) {
    return fakerKeywordMapper.array(current.args.items.map((schema) => parse(current, schema, { ...options, canOverride: false })).filter(Boolean))
  }

  if (isKeyword(current, schemaKeywords.enum)) {
    return fakerKeywordMapper.enum(
      current.args.items.map((schema) => {
        if (schema.format === 'number') {
          return schema.name
        }
        return transformers.stringify(schema.name)
      }),
    )
  }

  if (isKeyword(current, schemaKeywords.ref)) {
    if (!current.args?.name) {
      throw new Error(`Name not defined for keyword ${current.keyword}`)
    }

    if (options.canOverride) {
      return `${current.args.name}(data)`
    }

    return `${current.args.name}()`
  }

  if (isKeyword(current, schemaKeywords.object)) {
    const argsObject = Object.entries(current.args?.properties || {})
      .filter((item) => {
        const schema = item[1]
        return schema && typeof schema.map === 'function'
      })
      .map(([name, schemas]) => {
        const nameSchema = schemas.find((schema) => schema.keyword === schemaKeywords.name) as SchemaKeywordMapper['name']
        const mappedName = nameSchema?.args || name

        // custom mapper(pluginOptions)
        if (options.mapper?.[mappedName]) {
          return `"${name}": ${options.mapper?.[mappedName]}`
        }

        return `"${name}": ${joinItems(
          schemas
            .sort(schemaKeywordsorter)
            .map((schema) => parse(current, schema, { ...options, canOverride: false }))
            .filter(Boolean),
        )}`
      })
      .join(',')

    return `{${argsObject}}`
  }

  if (isKeyword(current, schemaKeywords.tuple)) {
    if (Array.isArray(current.args.items)) {
      return fakerKeywordMapper.tuple(current.args.items.map((schema) => parse(current, schema, { ...options, canOverride: false })).filter(Boolean))
    }

    return parse(current, current.args.items, { ...options, canOverride: false })
  }

  if (isKeyword(current, schemaKeywords.const)) {
    if (current.args.format === 'number' && current.args.name !== undefined) {
      return fakerKeywordMapper.const(current.args.name?.toString())
    }
    return fakerKeywordMapper.const(transformers.stringify(current.args.value))
  }

  if (isKeyword(current, schemaKeywords.matches) && current.args) {
    return fakerKeywordMapper.matches(current.args, options.regexGenerator)
  }

  if (isKeyword(current, schemaKeywords.null) || isKeyword(current, schemaKeywords.undefined) || isKeyword(current, schemaKeywords.any)) {
    return value() || ''
  }

  if (isKeyword(current, schemaKeywords.string)) {
    if (parent) {
      const minSchema = SchemaGenerator.find([parent], schemaKeywords.min)
      const maxSchema = SchemaGenerator.find([parent], schemaKeywords.max)

      return fakerKeywordMapper.string(minSchema?.args, maxSchema?.args)
    }

    return fakerKeywordMapper.string()
  }

  if (isKeyword(current, schemaKeywords.number)) {
    if (parent) {
      const minSchema = SchemaGenerator.find([parent], schemaKeywords.min)
      const maxSchema = SchemaGenerator.find([parent], schemaKeywords.max)

      return fakerKeywordMapper.number(minSchema?.args, maxSchema?.args)
    }

    return fakerKeywordMapper.number()
  }

  if (isKeyword(current, schemaKeywords.integer)) {
    if (parent) {
      const minSchema = SchemaGenerator.find([parent], schemaKeywords.min)
      const maxSchema = SchemaGenerator.find([parent], schemaKeywords.max)

      return fakerKeywordMapper.integer(minSchema?.args, maxSchema?.args)
    }

    return fakerKeywordMapper.integer()
  }

  if (isKeyword(current, schemaKeywords.datetime)) {
    return fakerKeywordMapper.datetime()
  }

  if (isKeyword(current, schemaKeywords.date)) {
    return fakerKeywordMapper.date(current.args.type, options.dateParser)
  }

  if (isKeyword(current, schemaKeywords.time)) {
    return fakerKeywordMapper.time(current.args.type, options.dateParser)
  }

  if (current.keyword in fakerKeywordMapper && 'args' in current) {
    const value = fakerKeywordMapper[current.keyword as keyof typeof fakerKeywordMapper] as (typeof fakerKeywordMapper)['const']

    const options = JSON.stringify((current as SchemaKeywordBase<unknown>).args)

    return value(options)
  }

  if (current.keyword in fakerKeywordMapper) {
    return value()
  }

  return undefined
}
