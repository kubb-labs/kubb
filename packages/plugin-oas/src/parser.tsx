import type { PluginFactoryOptions } from '@kubb/core'
import type { Operation, SchemaObject } from '@kubb/oas'
import { App, createRoot, type KubbNode } from '@kubb/react'
import type * as KubbFile from '@kubb/fs/types'
import type { OperationsByMethod } from './types.ts'
import { Oas } from '@kubb/plugin-oas/components'
import type { OperationGenerator } from './OperationGenerator.ts'
import type { SchemaGenerator, SchemaGeneratorOptions } from './SchemaGenerator.ts'

type OperationsProps<TOptions extends PluginFactoryOptions> = {
  generator: OperationGenerator<TOptions>
  options: TOptions['resolvedOptions']
  operations: Array<Operation>
  operationsByMethod: OperationsByMethod
}

type OperationProps<TOptions extends PluginFactoryOptions> = {
  generator: OperationGenerator<TOptions>
  options: TOptions['resolvedOptions']
  operation: Operation
}

type SchemaProps<TOptions extends PluginFactoryOptions> = {
  generator: SchemaGenerator<SchemaGeneratorOptions, TOptions>
  name: string
  schema: SchemaObject
  options: TOptions['resolvedOptions']
}

export type ParserOptions<TOptions extends PluginFactoryOptions> = {
  name: string
  operations?: (props: OperationsProps<TOptions>) => Promise<KubbFile.File[]>
  operation?: (props: OperationProps<TOptions>) => Promise<KubbFile.File[]>
  schema?: (props: SchemaProps<TOptions>) => Promise<KubbFile.File[]>
}

export type Parser<TOptions extends PluginFactoryOptions> = ParserOptions<TOptions>

export function createParser<TOptions extends PluginFactoryOptions>(parseOptions: ParserOptions<TOptions>): Parser<TOptions> {
  return parseOptions
}

export type ParserReactOptions<TOptions extends PluginFactoryOptions> = {
  name: string
  Operations?: (props: OperationsProps<TOptions>) => KubbNode
  Operation?: (props: OperationProps<TOptions>) => KubbNode
  Schema?: (props: SchemaProps<TOptions>) => KubbNode
}

export function createReactParser<TOptions extends PluginFactoryOptions>(parseOptions: ParserReactOptions<TOptions>): Parser<TOptions> {
  return {
    ...parseOptions,
    async operations({ generator, options, operations, operationsByMethod }) {
      if (!parseOptions.Operations) {
        return []
      }

      const { pluginManager, oas, plugin, mode } = generator.context
      const root = createRoot({
        logger: pluginManager.logger,
      })

      root.render(
        <App pluginManager={pluginManager} plugin={plugin} mode={mode}>
          <Oas oas={oas} operations={operations} generator={generator}>
            <parseOptions.Operations operations={operations} generator={generator} operationsByMethod={operationsByMethod} options={options} />
          </Oas>
        </App>,
      )

      return root.files
    },
    async operation({ generator, operation, options }) {
      if (!parseOptions.Operation) {
        return []
      }

      const { pluginManager, oas, plugin, mode } = generator.context
      const root = createRoot({
        logger: pluginManager.logger,
      })

      root.render(
        <App pluginManager={pluginManager} plugin={{ ...plugin, options }} mode={mode}>
          <Oas oas={oas} operations={[operation]} generator={generator}>
            <Oas.Operation operation={operation}>
              <parseOptions.Operation operation={operation} options={options} generator={generator} />
            </Oas.Operation>
          </Oas>
        </App>,
      )

      return root.files
    },
    async schema({ generator, schema, name, options }) {
      if (!parseOptions.Schema) {
        return []
      }

      const { pluginManager, oas, plugin, mode } = generator.context
      const root = createRoot({
        logger: pluginManager.logger,
      })

      const tree = generator.parse({ schema, name })

      root.render(
        <App pluginManager={pluginManager} plugin={{ ...plugin, options }} mode={mode}>
          <Oas oas={oas}>
            <Oas.Schema name={name} value={schema} tree={tree}>
              <parseOptions.Schema schema={schema} options={options} generator={generator} name={name} />
            </Oas.Schema>
          </Oas>
        </App>,
      )

      return root.files
    },
  }
}
