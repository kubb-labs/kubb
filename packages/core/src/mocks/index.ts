import path from 'node:path'
import { getSource } from '../FileManager'
import type { File, ResolvedFile } from '../fs/types.ts'
import type { Logger } from '../logger'
import type { PluginManager } from '../PluginManager.ts'
import { camelCase, pascalCase } from '../transformers/casing.ts'
import type { Plugin } from '../types.ts'
import { createFile } from '../utils'

export const mockedLogger = {
  emit(_type, _message) {},
  on(_type, _message) {},
  consola: {},
} as Logger

export const createMockedPluginManager = (name?: string) =>
  ({
    resolveName: (result) => {
      if (result.type === 'file') {
        return camelCase(name || result.name)
      }

      if (result.type === 'type') {
        return pascalCase(result.name)
      }

      if (result.type === 'function') {
        return camelCase(result.name)
      }

      return camelCase(result.name)
    },
    config: {
      output: {
        path: './path',
      },
    },
    resolvePath: ({ baseName }) => baseName,
    logger: {
      emit(message) {
        console.log(message)
      },
      on(_eventName, _args) {},
      logLevel: 3,
    },
    getPluginByKey: (_pluginKey: Plugin['key']) => {
      return undefined
    },
    getFile: ({ name, extname, pluginKey }) => {
      const baseName = `${name}${extname}`

      return {
        path: baseName,
        baseName,
        meta: {
          pluginKey,
        },
      }
    },
  }) as PluginManager

export const mockedPluginManager = createMockedPluginManager('')

export async function matchFiles(files: Array<ResolvedFile | File> | undefined, pre?: string) {
  if (!files) {
    return undefined
  }

  for await (const file of files) {
    const source = await getSource(createFile(file), { logger: mockedLogger })
    await expect(source).toMatchFileSnapshot(path.join(...['__snapshots__', pre, file.path].filter(Boolean)))
  }
}
