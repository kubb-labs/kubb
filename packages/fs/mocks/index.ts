import { pascalCase } from '../src/transformers/casing.ts'

import { FileManager, type KubbFile } from '../src'
import type { PluginManager } from '../src/PluginManager.ts'
import { readSync } from '../src/fs/read.ts'

export const mockedPluginManager = {
  resolveName: ({ name, type }) => {
    if (type === 'type') {
      return pascalCase(name)
    }

    return name
  },
  resolvePath: ({ baseName }) => baseName,
  logger: {
    emit(message) {
      console.log(message)
    },
    on(eventName, args) {},
    logLevel: 'info',
  },
  getFile: ({ name, extname, pluginKey }) => {
    const baseName = `${name}${extname}`
    let source = ''

    try {
      source = readSync(baseName)
    } catch (_e) {
      //
    }

    return {
      path: baseName,
      baseName,
      meta: {
        pluginKey,
      },
      source,
    }
  },
} as PluginManager

export async function matchFiles(files: KubbFile.File[]) {
  for (const file of files) {
    const source = await FileManager.getSource(file)
    expect(source).toMatchSnapshot()
  }
}
