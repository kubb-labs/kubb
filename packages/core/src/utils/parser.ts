import { extname } from 'node:path'
import type * as KubbFile from '@kubb/fs/types'

import { getRelativePath } from '@kubb/fs'
import hash from 'object-hash'
import { combineExports, combineImports, combineSources } from '../FileManager.ts'
import type { Logger } from '../logger.ts'

/**
 * Helper to create a file with name and id set
 */
export function createFile<TMeta extends object = object>(file: KubbFile.File<TMeta>): KubbFile.ResolvedFile<TMeta> {
  const extName = extname(file.baseName) as KubbFile.Extname

  if (!extName) {
    throw new Error(`No extName found for ${file.baseName}`)
  }

  const source = file.sources.map((item) => item.value).join('\n\n')
  const exports = file.exports ? combineExports(file.exports) : []
  const imports = file.imports && source ? combineImports(file.imports, exports, source) : []
  const sources = file.sources ? combineSources(file.sources) : []

  return {
    ...file,
    id: hash({ path: file.path }),
    name: trimExtName(file.baseName),
    extName,
    imports: imports.map((item) => createFileImport(item)),
    exports: exports.map((item) => createFileExport(item)),
    sources: sources.map((item) => createFileSource(item)),
  }
}

/**
 * Helper to create a fileImport with extName set
 */
export function createFileSource(source: KubbFile.Source): KubbFile.Source {
  return source
}

/**
 * Helper to create a fileImport with extName set
 */
export function createFileImport(imp: KubbFile.Import): KubbFile.ResolvedImport {
  const extName = extname(imp.path) as KubbFile.Extname

  return {
    ...imp,
    extName: imp.extName ? imp.extName : extName,
  }
}

/**
 * Helper to create a fileExport with extName set
 */
export function createFileExport(exp: KubbFile.Export): KubbFile.ResolvedExport {
  const extName = extname(exp.path) as KubbFile.Extname

  return {
    ...exp,
    extName: exp.extName ? exp.extName : extName,
  }
}

export type ParserModule<TMeta extends object = object> = {
  /**
   * By default @kubb/react is used
   */
  render: (item: any) => any
  /**
   * Convert a file to string
   */
  print: (file: KubbFile.ResolvedFile<TMeta>, options: PrintOptions) => Promise<string>
}

export function createFileParser<TMeta extends object = object>(parser: ParserModule<TMeta>): ParserModule<TMeta> {
  return parser
}

type PrintOptions = {
  logger?: Logger
}

const typeScriptParser = createFileParser({
  render() {
    return undefined
  },
  async print(file) {
    const module = await import('@kubb/parser-ts')

    const source = file.sources.map((item) => item.value).join('\n\n')

    const importNodes = file.imports
      .map((item) => {
        const path = item.root ? getRelativePath(item.root, item.path) : item.path

        return module.factory.createImportDeclaration({
          name: item.name,
          path,
          isTypeOnly: item.isTypeOnly,
        })
      })
      .filter(Boolean)

    const exportNodes = file.exports
      .map((item) => {
        return module.factory.createExportDeclaration({
          name: item.name,
          path: item.path,
          isTypeOnly: item.isTypeOnly,
          asAlias: item.asAlias,
        })
      })
      .filter(Boolean)

    // do some basic linting with the ts compiler
    return module.print([], { source: [module.print([...importNodes, ...exportNodes]), source].join('\n'), noEmitHelpers: false })
  },
})

const defaultParser = createFileParser({
  render() {
    return undefined
  },
  async print(file, { logger }) {
    return file.sources.map((item) => item.value).join('\n\n')
  },
})

const parsers: Record<KubbFile.Extname, ParserModule<any>> = {
  '.ts': typeScriptParser,
  '.js': typeScriptParser,
  '.jsx': typeScriptParser,
  '.tsx': typeScriptParser,
  '.json': defaultParser,
}

export async function getFileParser<TMeta extends object = object>(extName: KubbFile.Extname | undefined): Promise<ParserModule<TMeta>> {
  if (!extName) {
    return defaultParser
  }

  const parser = parsers[extName]

  if (!parser) {
    console.warn(`[parser] No parser found for ${extName}, default parser will be used`)
  }

  return parser || defaultParser
}

export function trimExtName(text: string): string {
  const extName = text.split('.').pop()

  return text.replace(`.${extName}`, '')
}
