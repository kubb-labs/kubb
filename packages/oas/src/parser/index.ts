import OASNormalize from 'oas-normalize'
import swagger2openapi from 'swagger2openapi'

import { Oas } from '../Oas.ts'
import { filterAndSort, isOpenApiV2Document } from '../utils.ts'

import type { OASDocument } from 'oas/types'
import type { OpenAPI } from 'openapi-types'

export type FormatOptions = {
  verbose?: boolean
  'no-sort'?: boolean
  sort?: boolean
  output?: string
  sortSet?: {
    root?: Array<'openapi' | 'info' | 'servers' | 'paths' | 'components' | 'tags' | 'x-tagGroups' | 'externalDocs'>
    get?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    post?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    put?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    patch?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    delete?: Array<'operationId' | 'summary' | 'description' | 'parameters' | 'requestBody' | 'responses'>
    parameters?: Array<'name' | 'in' | 'description' | 'required' | 'schema'>
    requestBody?: Array<'description' | 'required' | 'content'>
    responses?: Array<'description' | 'headers' | 'content' | 'links'>
    content?: Array<string>
    components?: Array<'parameters' | 'schemas'>
    schema?: Array<'description' | 'type' | 'items' | 'properties' | 'format' | 'example' | 'default'>
    schemas?: Array<'description' | 'type' | 'items' | 'properties' | 'format' | 'example' | 'default'>
    properties?: Array<'description' | 'type' | 'items' | 'format' | 'example' | 'default' | 'enum'>
  }
  filterSet?: {
    methods?: Array<'get' | 'post' | 'put' | 'patch' | 'delete' | 'options' | 'trace' | 'head' | 'parameters'>
    inverseMethods?: Array<'get' | 'post' | 'put' | 'patch' | 'delete' | 'options' | 'trace' | 'head' | 'parameters'>
    tags?: Array<string>
    inverseTags?: Array<string>
    operationIds?: Array<string>
    inverseOperationIds?: Array<string>
    operations?: Array<string>
    flags?: Array<string>
    inverseFlags?: Array<string>
    flagValues?: Array<string>
    inverseFlagValues?: Array<string>
    stripFlags?: Array<string>
    responseContent?: Array<string>
    inverseResponseContent?: Array<string>
    unusedComponents?: Array<'requestBodies' | 'schemas' | 'parameters' | 'responses'>
  }
  sortComponentsSet?: {}
  casingSet?: {}
}

export async function parse(pathOrApi: string | OASDocument, options: FormatOptions = {}): Promise<Oas> {
  const oasNormalize = new OASNormalize(pathOrApi, {
    enablePaths: true,
    colorizeErrors: true,
  })
  let document: OpenAPI.Document

  try {
    // resolve external refs
    document = await oasNormalize.bundle()
  } catch (e) {
    document = (await oasNormalize.load()) as OpenAPI.Document
  }

  if (isOpenApiV2Document(document)) {
    const { openapi } = await swagger2openapi.convertObj(document, {
      anchors: true,
    })

    const oas = await filterAndSort(openapi as OASDocument, options)

    return new Oas({ oas: oas as OASDocument })
  }

  const oas = await filterAndSort(document as OASDocument, options)

  return new Oas({ oas })
}
