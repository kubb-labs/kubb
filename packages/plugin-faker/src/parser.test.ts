import { schemas } from '@kubb/plugin-oas/mocks'

import * as parserFaker from './parser.ts'

describe('faker parse', () => {
  test.each(schemas.basic)('$name', ({ name, schema }) => {
    const text = parserFaker.parse({ parent: undefined, current: schema, siblings: [schema] }, { name, typeName: 'Pet' })
    expect(text).toMatchSnapshot()
  })
})
