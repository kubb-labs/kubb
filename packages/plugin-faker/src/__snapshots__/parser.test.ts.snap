// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`faker parse > 'and' 1`] = `"Object.assign({}, faker.string.alpha(), faker.number.float())"`;

exports[`faker parse > 'any' 1`] = `"undefined"`;

exports[`faker parse > 'array' 1`] = `"faker.helpers.multiple(() => (faker.helpers.arrayElement<any>([faker.number.float(), faker.string.alpha()])))"`;

exports[`faker parse > 'arrayAdvanced' 1`] = `"faker.helpers.multiple(() => (faker.number.float({ min: 1, max: 10 })), { count: { min: 3, max: 10 }})"`;

exports[`faker parse > 'arrayEmpty' 1`] = `"faker.helpers.multiple(() => (undefined))"`;

exports[`faker parse > 'arrayRef' 1`] = `"faker.helpers.multiple(() => (Pet()))"`;

exports[`faker parse > 'arrayRegex' 1`] = `"faker.helpers.multiple(() => (faker.helpers.fromRegExp("^[a-zA-Z0-9]{1,13}$")), { count: { min: 3, max: 10 }})"`;

exports[`faker parse > 'blob' 1`] = `"faker.image.url() as unknown as Blob"`;

exports[`faker parse > 'boolean' 1`] = `"faker.datatype.boolean()"`;

exports[`faker parse > 'catchall' 1`] = `"{}"`;

exports[`faker parse > 'const' 1`] = `""""`;

exports[`faker parse > 'date' 1`] = `"faker.date.anytime().toISOString().substring(0, 10)"`;

exports[`faker parse > 'datetime' 1`] = `"faker.date.anytime().toISOString()"`;

exports[`faker parse > 'default' 1`] = `undefined`;

exports[`faker parse > 'default' 2`] = `undefined`;

exports[`faker parse > 'enum' 1`] = `"faker.helpers.arrayElement<any>(["A", "B", "C", 2])"`;

exports[`faker parse > 'enumLiteralBoolean' 1`] = `"faker.helpers.arrayElement<any>([true, false])"`;

exports[`faker parse > 'integer' 1`] = `"faker.number.int()"`;

exports[`faker parse > 'matches' 1`] = `"faker.helpers.fromRegExp("^[A-Z]{2}$")"`;

exports[`faker parse > 'matchesReg' 1`] = `"faker.helpers.fromRegExp("/node_modules/")"`;

exports[`faker parse > 'max' 1`] = `undefined`;

exports[`faker parse > 'min' 1`] = `undefined`;

exports[`faker parse > 'nullable' 1`] = `undefined`;

exports[`faker parse > 'nullableAdditionalProperties' 1`] = `"{}"`;

exports[`faker parse > 'number' 1`] = `"faker.number.float()"`;

exports[`faker parse > 'object' 1`] = `"{"firstName": faker.string.alpha({ length: 2 }),"address": faker.string.alpha()}"`;

exports[`faker parse > 'objectAnd' 1`] = `"{"firstName": faker.string.alpha({ length: 2 }),"age": faker.number.float({ min: 3 }),"address": Object.assign({}, faker.string.alpha(), faker.number.float())}"`;

exports[`faker parse > 'objectArray' 1`] = `"{"ids": faker.helpers.multiple(() => (faker.helpers.fromRegExp("^[a-zA-Z0-9]{1,13}$")), { count: { min: 3, max: 10 }})}"`;

exports[`faker parse > 'objectArrayObject' 1`] = `"{"ids": faker.helpers.multiple(() => ({"enum": faker.helpers.arrayElement<NonNullable<NonNullable<NonNullable<Pet>["ids"]>[number]>["enum"]>(["A", "B", "C", 2])}), { count: { min: 3, max: 10 }})}"`;

exports[`faker parse > 'objectDates' 1`] = `"{"dateTime": faker.date.anytime().toISOString(),"date": faker.date.anytime().toISOString().substring(0, 10),"time": faker.date.anytime().toISOString().substring(11, 19),"nativeDate": faker.date.anytime()}"`;

exports[`faker parse > 'objectEmpty' 1`] = `"{}"`;

exports[`faker parse > 'objectEnum' 1`] = `"{"version": faker.helpers.arrayElement<NonNullable<Pet>["version"]>(["A", "B", "C", 2])}"`;

exports[`faker parse > 'objectObjectEnum' 1`] = `"{"prop1": {"prop2": faker.helpers.arrayElement<NonNullable<NonNullable<Pet>["prop1"]>["prop2"]>(["A", "B"])}}"`;

exports[`faker parse > 'objectOptional' 1`] = `"{"firstName": faker.string.alpha({ length: 2 }),"address": faker.string.alpha()}"`;

exports[`faker parse > 'primitiveDate' 1`] = `"faker.date.anytime()"`;

exports[`faker parse > 'ref' 1`] = `"Pet()"`;

exports[`faker parse > 'string' 1`] = `"faker.string.alpha()"`;

exports[`faker parse > 'stringLocal' 1`] = `"faker.date.anytime().toISOString()"`;

exports[`faker parse > 'stringOffset' 1`] = `"faker.date.anytime().toISOString()"`;

exports[`faker parse > 'time' 1`] = `"faker.date.anytime().toISOString().substring(11, 19)"`;

exports[`faker parse > 'tuple' 1`] = `"[]"`;

exports[`faker parse > 'tupleMulti' 1`] = `"[faker.string.alpha(), faker.number.float()]"`;

exports[`faker parse > 'undefined' 1`] = `"undefined"`;

exports[`faker parse > 'union' 1`] = `"faker.helpers.arrayElement<any>([faker.string.alpha(), faker.number.float()])"`;

exports[`faker parse > 'unionOne' 1`] = `"faker.helpers.arrayElement<any>([faker.string.alpha()])"`;

exports[`faker parse > 'unknown' 1`] = `"undefined"`;
